README.md
Neptune.io agent
=================
Neptune agent is a light weight GO agent which runs on your server to execute runbooks.

Agent installation:
-------------------
Agent can be installed with a one-liner. Please see http://docs.neptune.io/docs/agent-installation for more details.

Agent administration:
---------------------
Once you install agent you can administer it by running

    sudo service neptune-agentd [start|stop|restart|uninstall]

See http://docs.neptune.io/docs/agent-administration for detailed documentation.

Contributing
------------
Send a pull request to the master branch of this repo.

License and Authors
-------------------
Authors: Neptune.io, Inc.


ChangeLog.md
### Release 1.1.1
1. Compile with Go 1.6
2. Add arm (v6) and armv8 (64bit) support
3. Improve agentId checks


scripts/linux/mass_deployment.sh
#!/bin/sh

########################
# Script to run a command on multiple hosts through SSH ; Can be used to install Neptuneio agent and run any other workload scripts remotely
# Usage : ./mass_deployment.sh
# Description : This script uses AWS CLI to get a list of ec2 hostnames and runs workload scripts on all the hosts
# Requirements : AWS CLI - Make sure your aws config has access to call aws ec2 describe-instances
# Contact : support@neptune.io
########################

########################
# Section 1 : Change these variables as appropriate
########################

# SSH keypair directory & name
KEYPAIR_DIRECTORY=`eval echo ~/.ssh/keypairs`
KEYPAIR="keypair_name.pem"

# AWS Region
AWS_REGION="us-east-1"

# AWS Tag (All servers under a tag will get agent installed)
TAG_KEY="Name"
TAG_VALUE="NodeAppCluster"

# Alternatively use Opsworks stackID instead of AWS tag above.(All servers under a stackID will get agent installed)
OPSWORKS_STACKID="OPSWORKS_STACK_ID_HERE"

# Username for SSH (Usually for Amazon linux its ec2-user, for RedHat & SUSE its root , for Ubuntu its ubuntu )
USERNAME="ec2-user"

########################
# Section 2 : Don't change these variables
########################

SSH_OPTIONS="-o StrictHostKeyChecking=no"

#########################

#################
# Section 3 : Copy the agent installation command with your Neptune.io API key into AGENT_INSTALL_CMD variable
#################

AGENT_INSTALL_CMD='AGENT_USER="neptune" API_KEY="API_KEY_HERE" bash -c "$(curl -sS -L https://raw.githubusercontent.com/neptuneio/neptune-agent/prod/scripts/linux/install_neptune_agent_linux.sh)"'
#################
# Agent control commands : Don't Change
#################

AGENT_START_CMD='sudo service neptune-agentd start'
AGENT_STOP_CMD='sudo service neptune-agentd stop'
AGENT_UPDATE_CMD='sudo service neptune-agentd update'
AGENT_RESTART_CMD='sudo service neptune-agentd restart'

#################

#################
# Get a host list and install Neptune.io agent
#################

# Use AWS CLI descrbe-instances to get host list based on AWS TAGS

for host in `aws ec2 describe-instances --region $AWS_REGION --filters "Name=tag-key,Values=$TAG_KEY,Name=tag-value,Values=$TAG_VALUE" | grep PublicIpAddress | cut -d "\"" -f4` ; do

# Alternatively use AWS CLI descrbe-instances to get host list based on Opsworks stackID
# for host in `aws opsworks --region $AWS_REGION describe-instances --stack-id $OPSWORKS_STACKID | grep PublicIp | cut -d "\"" -f4` ; do

echo "Installing Neptuneio agent on $host ....\n"

################
# Comment or uncomment agent actions as needed
################

# Install agent
ssh $SSH_OPTIONS -i $KEYPAIR_DIRECTORY/$KEYPAIR -t $USERNAME@$host $AGENT_INSTALL_CMD

# Update and restart agent
# ssh $SSH_OPTIONS -i $KEYPAIR_DIRECTORY/$KEYPAIR -t $USERNAME@$host $AGENT_UPDATE_CMD

done


scripts/mac/install_nagent_osx.sh
#!/bin/bash

# Usage : Just run this script with root permissions
# Purpose : Install neptune agent and start it
# Steps :
# 1) Install dependencies including checking if python 2.6 is installed
# 2) Create an user for the agent to run as
# 3) Fetch latest and stable neptune agent (nagent.py)
# 4) Fetch agent plist to launch it as a service in OSX
# 5) Immediately start the agent as a particular user

# Global variables
NEPTUNEIO_AGENT="nagent.py"
DEFAULT_USER="neptuneioagent"
NAGENT_CONFIG="nagent.cfg"
NAGENT_PLIST="com.neptune.agent.plist"
STABLE_NAGENT_URL="https://raw.githubusercontent.com/neptuneio/nagent/prod/src"

# Output display colors
red='\033[0;31m'
green='\033[0;32m'
NC='\033[0m' # No Color

# Set the endpoint
if [ -n "$NEPTUNE_ENDPOINT" ]; then
    END_POINT="$NEPTUNE_ENDPOINT"
    STABLE_NAGENT_URL="https://raw.githubusercontent.com/neptuneio/nagent/staging/src"
else
    END_POINT="www.neptune.io"
fi

# If the user name is not given, take the neptuneio user.
if [ -z "$NAGENT_USER" ]; then
    NAGENT_USER="$DEFAULT_USER"
fi
echo "Username: $NAGENT_USER"

# Use curl or wget to download files
DOWNLOAD_CMD='curl -sS -o'
if which curl > /dev/null; then
    DOWNLOAD_CMD='curl -sS -o'
elif which wget > /dev/null; then
    DOWNLOAD_CMD='wget -q -O'
else
    echo 'No curl or wget found to download files ! Please install curl and rerun the command'
    exit 1
fi

# OS distribution should be Darwin, else prompt to user linux installer
OS_DISTRIBUTION=$(uname -s)
if [ "$OS_DISTRIBUTION" != "Darwin" ]; then
    echo -e "${red}Looks like this machine is not running on OSX. Please use linux installer instead ${NC}"
    exit 1
fi

# Check if python is installed
echo "-------------------------------------"
echo "Checking for python dependency"
if python -V ; then
    if `python -c "import sys; sys.exit(0 if sys.version_info < (2, 5) else 1)"`; then
        echo "Neptune agent needs Python >=2.6. Please upgrade Python and retry."y
        exit 1
    fi
    echo "Good to go..."
else
    echo -e "${red}Please install python 2.6 or higher and rerun. Exiting ${NC}"
    exit 1
fi

# Start install of agent
echo "Installing Neptuneio agent..."

# Install dependencies once based on linux distribution
if [ "$OS_DISTRIBUTION" == "Darwin" ]; then

    # Download pip directly and install
    $DOWNLOAD_CMD /tmp/get-pip.py https://bootstrap.pypa.io/get-pip.py
    sudo -H python /tmp/get-pip.py
    sudo rm -rf /tmp/get-pip.py

    # Install pip packages
    if which pip &> /dev/null; then
        PIP_CMD=`which pip`
    elif ls /usr/local/bin/pip &> /dev/null; then
        PIP_CMD='/usr/local/bin/pip'
    else
        echo -e "${red}Pip installation unsuccessful. Please install pip manually and retry${NC}"
        exit 1
    fi
    # sudo -H $PIP_CMD install -U pyopenssl ndg-httpsclient pyasn1
    sudo -H $PIP_CMD install -U simplejson
    sudo -H $PIP_CMD install -U boto
    # If python version is less than 2.7, install requests version 2.5.0 to avoid
    # some unnecessary warnings on console.
    if `python -c "import sys; sys.exit(0 if sys.version_info < (2, 7) else 1)"`; then
        sudo -H $PIP_CMD install requests==2.5.0
    else
        sudo -H $PIP_CMD install -U requests
    fi
fi

# Create the new user if it does not exist
sudo id -u $NAGENT_USER &>/dev/null
if [ $? -ne 0 ]; then
    sudo dscl . -create /Users/$NAGENT_USER
    sudo dscl . -create /Users/$NAGENT_USER UserShell /bin/bash
    sudo dscl . -create /Users/$NAGENT_USER RealName "Neptune agent"
    LAST_ID=`sudo dscl . -list /Users UniqueID | awk '{print $2}' | sort -n | tail -1`
    NEXT_ID=$((LAST_ID + 1))
    sudo dscl . -create /Users/$NAGENT_USER UniqueID $NEXT_ID
    sudo dscl . -create /Users/$NAGENT_USER PrimaryGroupID 20
    sudo dscl . -create /Users/$NAGENT_USER NFSHomeDirectory /Users/$NAGENT_USER
    sudo createhomedir -u $NAGENT_USER 2>&1 | grep -v "shell-init"
fi

# A Global variable but can be created only after creation of the user
NAGENT_HOME=`eval echo ~$NAGENT_USER/neptuneio`
echo "Home dir: $NAGENT_HOME"
sleep 2

# Create Neptune agent home directory
sudo mkdir -p $NAGENT_HOME

# Fetch the latest stable neptune agent and plist
echo "Fetching the latest stable version of neptuneio agent"
sudo $DOWNLOAD_CMD $NAGENT_HOME/$NEPTUNEIO_AGENT $STABLE_NAGENT_URL/$NEPTUNEIO_AGENT
sudo $DOWNLOAD_CMD $NAGENT_HOME/$NAGENT_PLIST $STABLE_NAGENT_URL/$NAGENT_PLIST

# Update user and agentn path in the plist
sudo sed -i.bak "s|AGENT_PATH|$NAGENT_HOME/$NEPTUNEIO_AGENT|" $NAGENT_HOME/$NAGENT_PLIST
sudo sed -i.bak "s|USER_NAME|$NAGENT_USER|" $NAGENT_HOME/$NAGENT_PLIST
sudo sed -i.bak "s|WORKING_DIRECTORY|$NAGENT_HOME|" $NAGENT_HOME/$NAGENT_PLIST
sudo rm -f $NAGENT_HOME/${NAGENT_PLIST}.bak

# Populate the neptuneio config
echo "Populating neptuneio agent config"
sudo bash -c "echo '[NEPTUNEIO]
API_KEY=$NEPTUNEIO_KEY
END_POINT=$END_POINT' > $NAGENT_HOME/$NAGENT_CONFIG"

# Add exec permissions to neptune agent, plist and config
sudo chmod 755 $NAGENT_HOME/$NEPTUNEIO_AGENT $NAGENT_HOME/$NAGENT_PLIST $NAGENT_HOME/$NAGENT_CONFIG

# Make the user as owner of the neptune agent home directory
sudo chown -R $NAGENT_USER $NAGENT_HOME

# Unload any existing neptune plist jobs
if [ -e  /Library/LaunchDaemons/$NAGENT_PLIST ]; then
    sudo launchctl unload -w /Library/LaunchDaemons/$NAGENT_PLIST
    sleep 3;
fi

# For OSX use launchctl to run it as a service
# Fetch agent plist
sudo cp $NAGENT_HOME/$NAGENT_PLIST /Library/LaunchDaemons

# Start the agent by loading the plist
echo "Starting Neptune agent..."
sudo launchctl load -w /Library/LaunchDaemons/$NAGENT_PLIST
# Check the status of daemon after 2 sec
sleep 2;
sudo launchctl list |grep neptune

echo "-------------------------------------"
echo "To check agent status run  : sudo launchctl list |grep neptune"
echo "To stop agent run          : sudo launchctl unload -w /Library/LaunchDaemons/$NAGENT_PLIST"
echo "To start agent run         : sudo launchctl load -w /Library/LaunchDaemons/$NAGENT_PLIST"
echo "Agent log available at     : $NAGENT_HOME/$NAGENT_LOG"



scripts/windows/installNeptuneAgent.ps1
$ErrorActionPreference="Stop"

# Global variables.
$arch = "$env:PROCESSOR_ARCHITECTURE".ToLower()
$agent = "neptune-agent.exe"
$config = "neptune-agent.json"
$stable_agent_url = "https://raw.githubusercontent.com/neptuneio/neptune-agent/prod/downloads/neptune-agent-windows-$arch.zip"
$api_key = $env:API_KEY
$endpoint = $env:END_POINT
$github_api_key = $env:GITHUB_API_KEY
$log_file_name = "neptune-agent.log"
$assigned_hostname = $env:ASSIGNED_HOSTNAME

# Install the agent in HOME_DIR\neptune
$INSTALL_PATH = Join-Path "$env:userprofile" "neptune"

# Function to download the artifacts
function Download-File {
    param ([string]$url,[string]$file)
    Write-Host "Downloading $url to $file"
    $downloader = new-object System.Net.WebClient
    $downloader.Proxy.Credentials=[System.Net.CredentialCache]::DefaultNetworkCredentials;
    $downloader.DownloadFile($url, $file)
}

# Helper function to unzip the artifacts
function Expand-ZIPFile($file, $destination) {
    $shell = new-object -com shell.application
    $zip = $shell.NameSpace($file)
    Write-Host "Extracting the zip file..."
    foreach($item in $zip.items()) {
        $shell.Namespace($destination).copyhere($item)
    }
}

if (!($api_key)) {
    Write-Error "API_KEY is mandatory. Please set it in the environment."
} else {
    Write-Host "Using the api key: $api_key"
}

if ($endpoint) {
    # Get the artifacts from master if the endpoint is specified.
    $stable_agent_url = "https://raw.githubusercontent.com/neptuneio/neptune-agent/master/downloads/neptune-agent-windows-$arch.zip"
} else {
    $endpoint = "www.neptune.io"
}

if (!($assigned_hostname)) {
    $assigned_hostname = ""
} else {
    Write-Host "Setting the assigned hostname to: $assigned_hostname"
}

Write-Host "Installing Neptune agent for windows ($arch)..."

if ($env:TEMP -eq $null) {
  $env:TEMP = Join-Path "$env:SystemDrive" "temp"
}
$temp = Join-Path "$env:TEMP" "neptune-install"

if (Test-Path $temp) {
    Remove-Item -Force -Recurse -Path $temp
}
New-Item -Force -ItemType directory -Path $temp > $null
$temp_zip_file = Join-Path "$temp" "neptune-agent-windows-$arch.zip"

# Download the artifacts
Download-File "$stable_agent_url" "$temp_zip_file"

# Unzip the artifacts in temp directory
Expand-ZIPFile "$temp_zip_file" "$temp"
if (!$?) {
   Write-Error "Could not unzip the artifacts."
}

# Construct the config file with the passed API_KEY and other values.
$temp_config = Join-Path $temp $config
(Get-Content $temp_config) -replace 'API_KEY_HERE', "$api_key" | Set-Content $temp_config
(Get-Content $temp_config) -replace 'END_POINT_HERE', "$endpoint" | Set-Content $temp_config
(Get-Content $temp_config) -replace 'AGENT_LOG_HERE', "$log_file_name" | Set-Content $temp_config
(Get-Content $temp_config) -replace 'ASSIGNED_HOSTNAME_HERE', "$assigned_hostname" | Set-Content $temp_config
(Get-Content $temp_config) -replace 'GITHUB_KEY_HERE', "$github_api_key" | Set-Content $temp_config

$agent_fullpath = Join-Path $INSTALL_PATH $agent
$config_fullpath = Join-Path $INSTALL_PATH $config
$public_cert_fullpath = Join-Path $INSTALL_PATH "neptuneio.crt"

if (Test-Path $INSTALL_PATH) {
    if (Test-Path $agent_fullpath) {
        # There might be an old instance of agent running already. Try to uninstall it.
        Write-Host "Stopping the old agent.."
        & $agent_fullpath stop

        Write-Host "Uninstalling the old agent service.."
        & $agent_fullpath uninstall
    }

    Remove-Item -Force -Recurse -Path $INSTALL_PATH
}

New-Item -Force -ItemType directory -Path $INSTALL_PATH > $null

# Copy the required files to install directory
Copy-Item -Force "$temp\$agent" $agent_fullpath
Copy-Item -Force "$temp\$config" $config_fullpath
Copy-Item -Force "$temp\neptuneio.crt" $public_cert_fullpath

# Install the agent with install command. This installs agent as NeptuneAgent service.
Write-Host "Installing the agent as service."
& $agent_fullpath install
if (!$?) {
    Write-Error "Could not install the agent as service."
} else {
    Write-Host "Successfully installed the agent as NeptuneAgent service. Now starting the service.."
}

# Start the agent service now.
& $agent_fullpath start
if (!$?) {
    Write-Error "Could not start the agent service."
} else {
    Write-Host "Successfully started the agent.."
}



scripts/linux/install_neptune_agent_linux.sh
#!/bin/bash

# Usage : Just run this script with root permissions
# Purpose : Installs neptune agent on a linux machine and starts it
# Steps :
# 1) Creates a new user called "neptune" or an user specified on the command line on the host
# 2) Fetches latest and stable neptune agent binary, config file and daemon
# 3) Push daemon to init.d (Currently supports sysV based init only)
# 4) Starts the agent as the specified user in step 1

# Global variables
NEPTUNE_AGENT_URL="https://raw.githubusercontent.com/neptuneio/neptune-agent/prod"
NEPTUNE_AGENT="neptune-agent"
NEPTUNE_AGENT_USER="neptune"
NEPTUNE_AGENT_DIR="agent"
NEPTUNE_AGENT_DAEMON="neptune-agentd"
NEPTUNE_AGENT_CONFIG="neptune-agent.json"
NEPTUNE_AGENT_LOG="neptune-agent.log"
DEFAULT_REQUIRE_SUDO="false"
NEPTUNE_END_POINT="www.neptune.io"
HOST_NAME=""
GITHUB_API_KEY=""
NEPTUNE_OLD_AGENT_DAEMON="nagentd"

# Output display colors
red='\033[0;31m'
green='\033[0;32m'
NC='\033[0m' # No Color

# Set the endpoint
if [ -n "$END_POINT" ]; then
    NEPTUNE_END_POINT="$END_POINT"
    # If endpoint is specified use master version of agent
    NEPTUNE_AGENT_URL="https://raw.githubusercontent.com/neptuneio/neptune-agent/master"
fi

# If the user name is specified on commandline, use it
if [ -n "$AGENT_USER" ]; then
    NEPTUNE_AGENT_USER="$AGENT_USER"
fi
echo "Username: $NEPTUNE_AGENT_USER"

# Check if user be given sudo permissions
if [ -z "$REQUIRE_SUDO" ]; then
    REQUIRE_SUDO="$DEFAULT_REQUIRE_SUDO"
fi
echo "Sudo priveleges : $REQUIRE_SUDO"

# Check if proper API_KEY is given, else exit
if [ -z "$API_KEY" ]; then
    echo -e "${red}Please give a proper API_KEY and retry installing agent. ${NC}"
    exit 1
fi

# Check if a hostname is assigned
if [ -n "$ASSIGNED_HOST_NAME" ]; then
    HOST_NAME=$ASSIGNED_HOST_NAME
fi

# Use curl or wget to download files
DOWNLOAD_CMD='curl -sS -L -o'
if which curl > /dev/null; then
    DOWNLOAD_CMD='curl -sS -L -o'
elif which wget > /dev/null; then
    DOWNLOAD_CMD='wget -q -O'
else
    echo -e "${red}No curl or wget found to download files ! Please install curl and rerun the command ${NC}"
    exit 1
fi

# Find the linux distribution
if [ -f /etc/redhat-release ];then
    # Start with default linux distro of Redhat
    LINUX_DISTRIBUTION="Redhat"
else
    LINUX_DISTRIBUTION=$(grep -Eo "(Debian|Ubuntu|CentOS|SUSE|Amazon)" /etc/issue 2>/dev/null)
fi

# Find the linux host architecture
UNAME=`uname -sp | awk '{print tolower($0)}'`

if [[ ($UNAME == *"mac os x"*) || ($UNAME == *darwin*) ]]; then
    PLATFORM="darwin"
    echo -e "${red}Please use mac or OSX installation script ${NC}"
    exit 1
elif [[ ($UNAME == *"freebsd"*) ]]; then
    PLATFORM="freebsd"
else
    PLATFORM="linux"
fi

case $UNAME in
    *x86_64*) ARCH="amd64" ;;
    *armv8*)  ARCH="arm64" ;;
    *arm*)    ARCH="arm"   ;;
    *)        ARCH="386"   ;;
esac


# Start install of agent
echo "Installing Neptune agent on $PLATFORM for $ARCH ..."

# Remove existing agents if any
if [ -e /etc/init.d/$NEPTUNE_AGENT_DAEMON ]; then
    sudo service $NEPTUNE_AGENT_DAEMON uninstall
fi
# Remove old python agent if it exists
if [ -e /etc/init.d/$NEPTUNE_OLD_AGENT_DAEMON ]; then
    sudo service $NEPTUNE_OLD_AGENT_DAEMON uninstall
fi
sleep 2

# Create the new user if it does not exist
sudo id -u $NEPTUNE_AGENT_USER &>/dev/null || sudo useradd $NEPTUNE_AGENT_USER -m

# A Global variable but can be created only after creation of the user
NEPTUNE_AGENT_HOME=`eval echo ~$NEPTUNE_AGENT_USER/$NEPTUNE_AGENT_DIR`
echo "Home dir: $NEPTUNE_AGENT_HOME"
sleep 2

# Create Neptune agent directory in user's home
sudo mkdir -p $NEPTUNE_AGENT_HOME
sleep 2

# Fetch the latest stable neptune agent daemon
if sudo bash -c "[ -e $NEPTUNE_AGENT_HOME/${NEPTUNE_AGENT}-${PLATFORM}-${ARCH}.tar.gz ]" && sudo bash -c "[ -e $NEPTUNE_AGENT_HOME/$NEPTUNE_AGENT_DAEMON ]"
then
    echo "Looks like you already downloaded agent binary and daemon.Using them..."
else
    # Download agent and daemon
    echo "Fetching the latest version of neptune agent and daemon"
    sudo $DOWNLOAD_CMD $NEPTUNE_AGENT_HOME/${NEPTUNE_AGENT}-${PLATFORM}-${ARCH}.tar.gz $NEPTUNE_AGENT_URL/downloads/${NEPTUNE_AGENT}-${PLATFORM}-${ARCH}.tar.gz
    sudo $DOWNLOAD_CMD $NEPTUNE_AGENT_HOME/$NEPTUNE_AGENT_DAEMON $NEPTUNE_AGENT_URL/scripts/$PLATFORM/$NEPTUNE_AGENT_DAEMON
fi

echo "Unzipping agent binaries"
sudo tar -zxf $NEPTUNE_AGENT_HOME/${NEPTUNE_AGENT}-${PLATFORM}-${ARCH}.tar.gz -C $NEPTUNE_AGENT_HOME

# Remove tar file if unzip is successful
if [ $? -eq 0 ] && sudo bash -c "[ -e $NEPTUNE_AGENT_HOME/$NEPTUNE_AGENT_DAEMON ]"
then
    sudo rm -rf $NEPTUNE_AGENT_HOME/${NEPTUNE_AGENT}-${PLATFORM}-${ARCH}.tar.gz
else
    echo -e "${red}Agent binary download didn't finish properly ${NC}"
    echo -e "${red}Please download $NEPTUNE_AGENT_URL/downloads/${NEPTUNE_AGENT}-${PLATFORM}-${ARCH}.tar.gz and put it in $NEPTUNE_AGENT_HOME ${NC}"
    echo -e "${red}Please download $NEPTUNE_AGENT_URL/scripts/$PLATFORM/$NEPTUNE_AGENT_DAEMON and put it in $NEPTUNE_AGENT_HOME"
    echo -e "${red}Re-run the installation command again. Exiting for now ${NC}"
    exit 1
fi

# Update repo URL in the daemon to enable agent updates
sudo sed -i "s|AGENT_USER_HERE|$NEPTUNE_AGENT_USER|" $NEPTUNE_AGENT_HOME/$NEPTUNE_AGENT_DAEMON

# Populate the neptune config
echo "Updating agent config"
sudo sed -i "s|API_KEY_HERE|$API_KEY|; s|END_POINT_HERE|$NEPTUNE_END_POINT|; s|AGENT_LOG_HERE|$NEPTUNE_AGENT_LOG|; s|ASSIGNED_HOSTNAME_HERE|$HOST_NAME|; s|GITHUB_KEY_HERE|$GITHUB_API_KEY|" $NEPTUNE_AGENT_HOME/$NEPTUNE_AGENT_CONFIG

# Add agent user to sudoers list and turn off requiretty
if [ "$REQUIRE_SUDO" == "true" ] || [ "$REQUIRE_SUDO" == "TRUE" ] || [ "$REQUIRE_SUDO" == "True" ]; then
    # When /etc/sudoers.d exists, add a local file instead of modifying /etc/sudoers directly
    if [ -d "/etc/sudoers.d" ]; then
        echo "Adding $NEPTUNE_AGENT_USER to sudoers list without tty requirement"

        if [ ! -f /etc/sudoers.d/neptune_sudo_perms ]; then
            sudo echo "$NEPTUNE_AGENT_USER ALL=(ALL) NOPASSWD:ALL
Defaults:$NEPTUNE_AGENT_USER !requiretty" > /tmp/neptune_sudo_perms.bak

            # Check the syntax with strict mode (-s) of the neptune sudo permissions file to make sure it is correct.
            sudo visudo -csf /tmp/neptune_sudo_perms.bak
            if [ $? -eq 0 ]; then
                # Give 0440 permissions and then place the new sudoers file as local sudoers file if syntax is correct.
                sudo chmod 0440 /tmp/neptune_sudo_perms.bak
                sudo cp /tmp/neptune_sudo_perms.bak /etc/sudoers.d/neptune_sudo_perms
                sudo chmod 0440 /etc/sudoers.d/neptune_sudo_perms
            else
                echo "Couldn't add $NEPTUNE_AGENT_USER to sudoers list programmatically"
                echo "Please give sudo permission manually by following instructions at http://docs.neptune.io/docs/sudo-priveleges-control"
            fi
        fi
    else
        echo "Couldn't add $NEPTUNE_AGENT_USER to sudoers list programmatically as your host doesn't have /etc/sudoers.d directory to safely add sudo permissions"
        echo "Please give sudo permission manually by following instructions at http://docs.neptune.io/docs/sudo-priveleges-control"
    fi
fi

# Add exec permissions to neptune agent and daemon
sudo chmod 755 $NEPTUNE_AGENT_HOME $NEPTUNE_AGENT_HOME/$NEPTUNE_AGENT_DAEMON $NEPTUNE_AGENT_HOME/$NEPTUNE_AGENT_CONFIG

# Make the user as owner of the neptune agent home directory
sudo chown -R $NEPTUNE_AGENT_USER $NEPTUNE_AGENT_HOME

# Copy neptune agent daemon to init.d directory
sudo cp $NEPTUNE_AGENT_HOME/$NEPTUNE_AGENT_DAEMON /etc/init.d/

# Based on linux distribution use chkconfig or update-rc.d

if [ "$LINUX_DISTRIBUTION" == "Amazon" -o "$LINUX_DISTRIBUTION" == "Redhat" -o "$LINUX_DISTRIBUTION" == "CentOS" -o "$LINUX_DISTRIBUTION" == "SUSE" ]; then
    sudo chkconfig --add  $NEPTUNE_AGENT_DAEMON
    sudo chkconfig $NEPTUNE_AGENT_DAEMON on
else
    sudo update-rc.d $NEPTUNE_AGENT_DAEMON start 90 1 2 3 5 . stop 10 0 6 .
fi

# Start the agent daemon immediately ; If running, restart
sudo service $NEPTUNE_AGENT_DAEMON restart
